pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -59,34 +65,67 @@ func NewScheduler(kubeClient client.Client) *Scheduler {
 func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 	constraints := provisioner.Spec.Constraints.DeepCopy()
+	start := time.Now()
+
+	sort.Slice(pods, byCPUAndMemoryDescending(pods))
+	sort.Slice(instanceTypes, func(i, j int) bool {
+		return instanceTypes[i].Price() < instanceTypes[j].Price()
+	})
+
 	// Inject temporarily adds specific NodeSelectors to pods, which are then
 	// used by scheduling logic. This isn't strictly necessary, but is a useful
 	// trick to avoid passing topology decisions through the scheduling code. It
 	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
 	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
 		return nil, fmt.Errorf("injecting topology, %w", err)
 	}
-	return s.schedule(constraints, instanceTypes, pods), nil
-}
 
-// schedule separates pods into a list of TheoreticalNodes that contain the pods. All pods in each theoretical node
-// contain isomorphic scheduling constraints and can be deployed together on the same node, or multiple similar nodes if
-// the pods exceed one node's capacity.
-func (s *Scheduler) schedule(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*Node {
-	var nodes []*Node
+	nodeSet, err := NewNodeSet(ctx, constraints, s.kubeClient)
+	if err != nil {
+		return nil, fmt.Errorf("constructing nodeset, %w", err)
+	}
 
 	for _, pod := range pods {
 		isScheduled := false
-		for _, node := range nodes {
+		for _, node := range nodeSet.nodes {
 			if err := node.Compatible(pod); err == nil {
 				node.Add(pod)
 				isScheduled = true
 				break
 			}
 		}
 		if !isScheduled {
-			nodes = append(nodes, NewNode(constraints, instanceTypes, pod))
+			n, err := NewNode(constraints, nodeSet.daemons, instanceTypes, pod)
+			if err != nil {
+				logging.FromContext(ctx).With("pod", client.ObjectKeyFromObject(pod)).Errorf("scheduling pod, %s", err)
+			} else {
+				nodeSet.Add(n)
+			}
+		}
+	}
+	logging.FromContext(ctx).Infof("scheduled %d pods onto %d nodes in %s", len(pods), len(nodeSet.nodes), time.Since(start))
+	return nodeSet.nodes, nil
+}
+
+func byCPUAndMemoryDescending(pods []*v1.Pod) func(i int, j int) bool {
+	return func(i, j int) bool {
+		lhs := resources.RequestsForPods(pods[i])
+		rhs := resources.RequestsForPods(pods[j])
+
+		cpuCmp := resources.Cmp(lhs[v1.ResourceCPU], rhs[v1.ResourceCPU])
+		if cpuCmp < 0 {
+			// LHS has less CPU, so it should be sorted after
```

Comment:
Why comment here and not on memory?

Reply:
The CPU check follows the same pattern of lesser-value => sorts after the other, figured documenting it once was enough.